subsys/bluetooth/controller/ll_sw/ull_sched.c

```diff
@@ -4,43 +4,536 @@
+
+	win_offset_calc(conn, 0, ticks_to_offset_next,
+			conn->llcp_conn_param.interval_max, &offset_max,
+			(void *)conn->llcp_conn_param.pdu_win_offset0);
 }
 
 void ull_sched_mfy_win_offset_select(void *param)
 {
-	/* TODO: */
+#define OFFSET_S_MAX 6
+#define OFFSET_M_MAX 6
+	u16_t win_offset_m[OFFSET_M_MAX] = {0, };
+	u8_t offset_m_max = OFFSET_M_MAX;
+	struct ll_conn *conn = param;
+	u8_t offset_index_s = 0U;
+	u8_t has_offset_s = 0U;
+	u32_t ticks_to_offset;
+	u16_t win_offset_s;
+
+	ticks_to_offset = HAL_TICKER_US_TO_TICKS(conn->llcp_conn_param.offset0 *
+						 1250);
+
+	win_offset_calc(conn, 1, &ticks_to_offset,
+			conn->llcp_conn_param.interval_max, &offset_m_max,
+			(void *)win_offset_m);
+
+	while (offset_index_s < OFFSET_S_MAX) {
+		u8_t offset_index_m = 0U;
+
+		memcpy((u8_t *)&win_offset_s,
+		       ((u8_t *)&conn->llcp_conn_param.offset0 +
+			(sizeof(u16_t) * offset_index_s)), sizeof(u16_t));
+
+		while (offset_index_m < offset_m_max) {
+			if (win_offset_s != 0xffff) {
+				if (win_offset_s ==
+				    win_offset_m[offset_index_m]) {
+					break;
+				}
+
+				has_offset_s = 1U;
+			}
+
+			offset_index_m++;
+		}
+
+		if (offset_index_m < offset_m_max) {
+			break;
+		}
+
+		offset_index_s++;
+	}
+
+	if (offset_index_s < OFFSET_S_MAX) {
+		conn->llcp_cu.win_offset_us = win_offset_s * 1250;
+		memcpy(conn->llcp.conn_upd.pdu_win_offset, &win_offset_s,
+		       sizeof(u16_t));
+	} else if (!has_offset_s) {
+		conn->llcp_cu.win_offset_us = win_offset_m[0] * 1250;
+		memcpy(conn->llcp.conn_upd.pdu_win_offset, &win_offset_m[0],
+		       sizeof(u16_t));
+	} else {
+		struct pdu_data *pdu_ctrl_tx;
+
+		/* procedure request acked */
+		conn->llcp_ack = conn->llcp_req;
+
+		/* CPR request acked */
+		conn->llcp_conn_param.ack = conn->llcp_conn_param.req;
+
+		/* reset mutex */
+		ull_conn_upd_curr_reset();
+
+		/* send reject_ind_ext */
+		pdu_ctrl_tx = (void *)
+			((u8_t *)conn->llcp.conn_upd.pdu_win_offset -
+			 offsetof(struct pdu_data,
+				  llctrl.conn_update_ind.win_offset));
+		pdu_ctrl_tx->ll_id = PDU_DATA_LLID_CTRL;
+		pdu_ctrl_tx->len =
+			offsetof(struct pdu_data_llctrl, reject_ext_ind) +
+			sizeof(struct pdu_data_llctrl_reject_ext_ind);
+		pdu_ctrl_tx->llctrl.opcode =
+			PDU_DATA_LLCTRL_TYPE_REJECT_EXT_IND;
+		pdu_ctrl_tx->llctrl.reject_ext_ind.reject_opcode =
+			PDU_DATA_LLCTRL_TYPE_CONN_PARAM_REQ;
+		pdu_ctrl_tx->llctrl.reject_ext_ind.error_code =
+			BT_HCI_ERR_UNSUPP_LL_PARAM_VAL;
+	}
+#undef OFFSET_S_MAX
+#undef OFFSET_M_MAX
+}
+
+static void win_offset_calc(struct ll_conn *conn_curr, u8_t is_select,
+			    u32_t *ticks_to_offset_next, u16_t conn_interval,
+			    u8_t *offset_max, u8_t *win_offset)
+{
+	u32_t ticks_prepare_reduced = 0U;
+	u32_t ticks_to_expire_prev;
+	u32_t ticks_slot_abs_prev;
+	u32_t ticks_slot_abs = 0U;
+	u32_t ticks_anchor_prev;
+	u32_t ticks_to_expire;
+	u8_t ticker_id_other;
+	u8_t ticker_id_prev;
+	u32_t ticks_anchor;
+	u8_t offset_index;
+	u16_t _win_offset;
```

Comment:
easily confused with the win_offset parameter. Maybe use another name, instead of just differing by underscore?